pyquil/wavefunction.py

```diff
@@ -1,27 +1,96 @@
 """
+import struct
+import warnings
+
+import numpy as np
+from six import integer_types
+
+OCTETS_PER_DOUBLE_FLOAT = 8
+OCTETS_PER_COMPLEX_DOUBLE = 2 * OCTETS_PER_DOUBLE_FLOAT
 
 
 class Wavefunction(object):
 
-    def __init__(self, amplitude_vector):
+    def __init__(self, amplitude_vector, classical_memory=None):
         """
         Initializes a wavefunction
         :param amplitude_vector: A numpy array of complex amplitudes
+        :param classical_memory: Any classical memory associated with this wavefunction result
         """
         if len(amplitude_vector) == 0 or len(amplitude_vector) & (len(amplitude_vector) - 1) != 0:
             raise TypeError("Amplitude vector must have a length that is a power of two")
 
         self.amplitudes = amplitude_vector
+        self.classical_memory = classical_memory
+
+    @staticmethod
+    def ground(qubit_num):
+        """
+        Constructs the groundstate wavefunction for a given number of qubits.
+
+        :param int qubit_num:
+        :return: A Wavefunction in the ground state
+        :rtype: Wavefunction
+        """
+        container = [0] * (2**qubit_num)
+        container[0] = 1.0
+        return Wavefunction(container)
+
+    @staticmethod
+    def from_bit_packed_string(coef_string, classical_addresses):
+        """
+        From a bit packed string, unpacks to get the wavefunction and classical measurement results
+        :param coef_string:
+        :param classical_addresses:
+        :return:
+        """
+        num_octets = len(coef_string)
+        num_addresses = len(classical_addresses)
+        num_memory_octets = _round_to_next_multiple(num_addresses, 8) // 8
+        num_wavefunction_octets = num_octets - num_memory_octets
+
+        # Parse the classical memory
+        mem = []
+        for i in range(num_memory_octets):
+            octet = struct.unpack('B', coef_string[i:i + 1])[0]
+            mem.extend(_octet_bits(octet))
+
+        mem = mem[0:num_addresses]
+
+        # Parse the wavefunction
+        wf = np.zeros(num_wavefunction_octets // OCTETS_PER_COMPLEX_DOUBLE, dtype=np.cfloat)
+        for i, p in enumerate(range(num_memory_octets, num_octets, OCTETS_PER_COMPLEX_DOUBLE)):
+            re_be = coef_string[p: p + OCTETS_PER_DOUBLE_FLOAT]
+            im_be = coef_string[p + OCTETS_PER_DOUBLE_FLOAT: p + OCTETS_PER_COMPLEX_DOUBLE]
+            re = struct.unpack('>d', re_be)[0]
+            im = struct.unpack('>d', im_be)[0]
+            wf[i] = complex(re, im)
+
+        return Wavefunction(wf, mem)
 
     def __len__(self):
         return len(self.amplitudes).bit_length() - 1
 
     def __iter__(self):
+        warnings.warn("""
+Previously, qvm.wavefunction returned both classical memory and wavefunction
+as a pair. Now it just returns a Wavefunction object.
+You likely need to change this:
+    wf, _ = qvm.wavefunction(program, ...)
+To just this:
+    wf = qvm.wavefunction(program, ...)\n""", stacklevel=2)
```

Comment:
does wavefunction take `stacklevel` as a param?

Reply:
This is a little tough to parse, the error message that the user sees is just this:
```
You likely need to change this:
    wf, _ = qvm.wavefunction(program, ...)
To just this:
    wf = qvm.wavefunction(program, ...)\n
```

stacklevel=2 is a parameter to warnings.warn which tells Python to print the line number of the function that called this method, instead of always printing the line number within wavefunction.py where warnings was invoked.

Reply:
haha there we go, looks like my visual compiler was on the fritz a bit